src/gsl.rs

```diff
@@ -28,24 +28,34 @@ pub(crate) struct Guard(Option<Repr>);
 
 #[derive(Debug)]
 enum Repr {
-    Read(RwLockReadGuard<'static, ()>),
+    Read(Option<RwLockReadGuard<'static, ()>>),
     Write(RwLockWriteGuard<'static, ()>),
 }
 
+thread_local! {
+    static CACHE: Cell<Cache> = Cell::new(Cache::Read(0));
```

Comment:
:+1: for moving this up! "This module has ambient global state" is a pretty important bit of info, and it should be clearly visible!

Reply:
Yeah I thought about not doing this since it's not "relevant" to the PR itself, but since the cache is used a lot in the read and write functions I thought it'd be better to show its definition sooner.